Create Rule “reducing-design-debt/rule” (#10164)
SSWConsulting/SSW.Rules.Content: rules/reducing-design-debt/rule.md

---
type: rule
tips: ""

> title: "Do you know how to avoid design debt? "

Reviewer: ```suggestion
title: Do you know how to avoid design debt?
```

seoDescription: Making ad-hoc UI changes without a designer leads to design
  debt. Learn how to prevent inconsistencies and poor UX by following a better
  workflow.
uri: avoid-design-debt
authors:
  - title: ""
created: 2025-04-23T11:45:00.000Z

> archivedreason: ""

Reviewer: Hi @micsblank 

Can you add some related rules?
I thought about the "less is more", the "designers test please" and "tech debt"

guid: e0d0ea3b-9b9f-465a-8da8-54f6f7280790
---

It’s tempting to fix small UI issues on the fly — change a button size, adjust some spacing, or tweak a color. But those “quick wins” often turn into long-term losses, leading to a messy, inconsistent UI that confuses users and slows development.

<!--endintro-->

Design debt is like technical debt: shortcuts that seem efficient in the moment create chaos down the line. Without a shared system, visual inconsistencies multiply, developers second-guess design intent, and user experience suffers.

## What causes design debt?

It usually starts with innocent intentions:

* "Just added a quick icon"
* "Tightened the padding a bit"
* "Didn't want to bother design — it's small"

We’ve all done it. But enough of these add up fast. Suddenly, the product feels inconsistent, design is out of sync, and developers redo work they thought was already done.

## How to prevent design debt

### 1. Before you code

Ask yourself:

* Is this already in the design system?
* Will users see this change?
* Does it affect layout, spacing, interaction, or visuals?

> If the answer is “yes” or you’re not sure — involve design.

Reviewer: ```suggestion
If the answer is “yes” or you’re not sure; it involves design. So ask for a “Test Please” and tag a designer on the PBI before making changes.
```

> *Use a “Test Please” or tag a designer on the PBI before making changes.*

Reviewer: ```suggestion
```

### 2. Track it properly

Even small tweaks need visibility:

* Create a PBI for every UI change
* Tag it `needs-design` or `minor-UI` depending on the impact
* Let the PO and designer review and prioritise

### 3. Use the design system

Design systems are like component libraries — but for UI.

> * If the component exists, use it.

Reviewer: Hi @micsblank 

See https://www.ssw.com.au/rules/avoid-full-stops-in-bullet-point-lists

Then remove full stops from lists ;)

* If it doesn’t, don’t freestyle. Ask if a new pattern is needed.

Always check with design. No matter what.

> ## Examples – The SSW Mega Menu

Reviewer: ```suggestion
## Example – The SSW Mega Menu
```

::: greybox
This Menu has had small tweaks added over time: new icons, padding changes, label tweaks — all without a design review.
:::
::: bad

> ![Figure: Bad Example – The current Mega Menu has accumulated inconsistencies that break from the original design system](design-debt-bad-ex.png)

Reviewer: ```suggestion
![Figure: Bad example – The Mega Menu has evolved and accumulated inconsistencies that breaks the original design](design-debt-bad-ex.png)
```

:::

::: greybox
This is the original Menu design — consistent, aligned with the design system, and reviewed by design before implementation.
:::
::: good

> ![Figure: Good Example – The Mega Menu as originally designed, clean and consistent with the design system](design-debt-good-ex.png)

Reviewer: ```suggestion
![Figure: Good example – The Mega Menu is clean and consistent as originally designed](design-debt-good-ex.png)
```

:::

## More ways to prevent design debt

* Screenshot your change and post it in Slack before merging

> * Ask for a quick designer 👀 on spacing, alignment, and component use

Reviewer: Hi @micsblank 

This sentence sounds weird... Should it say "Ask for a quick test please from a designer" ?

* Loop in design early on bigger stuff (e.g. layout or feature changes)
* After merge, let design know if you created something reusable

> ## Why design debt matters

Reviewer: Hi @micsblank 

I think this section should come as the 2nd one, before "how to prevent"

---
type: rule
tips: ""
title: "Do you know how to avoid design debt? "
seoDescription: Making ad-hoc UI changes without a designer leads to design
  debt. Learn how to prevent inconsistencies and poor UX by following a better
  workflow.
uri: avoid-design-debt
authors:
  - title: ""
created: 2025-04-23T11:45:00.000Z
related:
  - less-is-more
  - request-a-test-please
  - technical-debt
archivedreason: ""
guid: e0d0ea3b-9b9f-465a-8da8-54f6f7280790
---

It’s tempting to fix small UI issues on the fly — change a button size, adjust some spacing, or tweak a color. But those “quick wins” often turn into long-term losses, leading to a messy, inconsistent UI that confuses users and slows development.

<!--endintro-->

Design debt is like technical debt: shortcuts that seem efficient in the moment create chaos down the line. Without a shared system, visual inconsistencies multiply, developers second-guess design intent, and user experience suffers.

## What causes design debt?

It usually starts with innocent intentions:

* "Just added a quick icon"
* "Tightened the padding a bit"
* "Didn't want to bother design — it's small"

> We’ve all done it. But enough of these add up fast. Suddenly, the product feels inconsistent, design is out of sync, and developers redo work they thought was already done.

Reviewer 2: ```suggestion
We’ve all done it. But enough of these add up fast. Before you know it, the product starts to feel inconsistent, design is out of sync, and developers redo work they thought was already done.
```

## Why design debt matters

### 🚨 Why it happens

* Rushed timelines or MVP mindset (“we’ll fix it later”)
* No shared design system
* Designers and developers working in silos
* Unclear product direction or pivots

### 📉 Why it’s a problem

* Hurts user trust and usability
* Makes the product feel messy or inconsistent
* Slows future development and design
* Causes rework and team friction

### 🧹 How to manage it

* Run regular UX audits and design reviews
* Maintain a living design system or component library
* Include UI refactoring in your roadmap
* Document design decisions with clear rationale

## How to prevent design debt

### 1. Before you code

Ask yourself:

> * Is this already in the design system?

Reviewer 2: @jaydenalchin @micsblank 

I think this question is a bit misleading?

If yes, they need a designer.
If no, does that mean they don't?

Reviewer: Agree with @bettybondoc 

Maybe change from:

> Ask yourself:
> 
> * Is this already in the design system?

to

> First check if what you want to change is already in the design system. If yes, you just need to follow it. If not, then ask yourself:
